specs/kvstore/WriterFencing.fizz

```diff
@@ -0,0 +1,145 @@
+    atomic func WriteManifest(new_manifest):
+        store = manifest_store
+        write_status = store.ConditionalWrite(new_manifest.id, new_manifest)
+        if write_status == SUCCESS:
+            self.state = "ManifestUpdated"
+            self.manifest = new_manifest
+        else:
+            self.state = "fenced"
+        return write_status
+
+    
+    #db.rs\db\open_with_fp_registry() does below
+    atomic func Open():
+        self.state = "initializing"
+        load_manifest_result = self.LoadLatestManifest()
+        if load_manifest_result != SUCCESS:
+            return
+        self.WriteFencingWal()
+
+    #flush.rs\DbInner\flush() is the only method that increments next_wal on write.
+    atomic func WriteWal():
+        store = wal_store
+        write_status = store.ConditionalWrite(self.next_wal, self.manifest.writer_epoch)
+        if write_status == SUCCESS:
+            self.next_wal = self.next_wal + 1
+        else:
+            self.state = "fenced"
+        return write_status
+
+    #db.rs\DbInner\fence_writers()
+    atomic func WriteFencingWal():
+        wal = wal_store.List()
+        if(len(wal) > 0):
+            self.next_wal = wal[len(wal) - 1] + 1
+        else:
+            self.next_wal = 1
+
+        store = manifest_store
+        while self.state == "ManifestUpdated":
+            write_status = self.WriteWal()
+            if write_status == SUCCESS:
+                self.state = "opened"
+                self.was_ever_open = 1
+                break;
+            else:
+                manifests = store.List()
+                recent_manifest = store.Read(manifests[len(manifests) - 1])
+                # if there is a new writer, set current to failed.  Otherwise, try with next wal.
+                if self.manifest.writer_epoch < recent_manifest.writer_epoch:
+                    self.state = "fenced"
+                    break
+                else:
+                    self.next_wal = self.next_wal + 1
+                
+action Init:
+    wal_store = ObjectStore()
+    manifest_store = ObjectStore()
+    writers = []
+    writer_ids = range(0, NUM_WRITERS)
+    for i in range(0,NUM_WRITERS):
+        writers.append(SlateDBWriter())
+
+atomic fair action WriterOpen:
+    writer_id = any writer_ids
+    current_writer = writers[writer_id]
+    if current_writer.state == "waiting":
+        current_writer.Open()
+
+always assertion NotMoreThanOneWriterInWritableState:
+    writable_writer_count = 0
+    wal_list = wal_store.List()
+    for i in range(0, NUM_WRITERS):
+        # writer state must be opened, and next wal slot be available in object store for the write to succeed.
+        if writers[i].state == "opened" and writers[i].next_wal not in wal_list:
+            writable_writer_count +=1
+    return writable_writer_count <=1
+
+# After a new writer writes wal, old writer cannot write its wal. Assert by reading writer epoch of sequential wal.
+always assertion OlderWriterCannotAddWal:
```

Comment:
This assert fails in the model checker with below error state.
![error-graph](https://github.com/user-attachments/assets/15ee6a2b-30d2-4d6b-9c85-288273fe36a7)

Thanks @jayaprabhakar  for suggesting to remove "atomic". I am still debugging to see if it is an actual issue or an issue with the fizzbee model.

Reply:
The case that it reports is

1. Writer 1 - creates manifest `1.manifest` with epoch 1
2. Writer 2 - creates manifest `2.manifest` with epoch 2
3. Writer 2 writes `wal/1.sst`. It is now open
4. Writer 1 lists wals, and creates `wal/2.sst`. It is also now open.
5. Writer 1 can now write wal, and it is not the owner.

Reply:
Oh this looks like an actual bug! I think this might be why we wanted to put the epoch in the object metadata. This way, when Writer 1 goes to list WALs and then replay them, it'll see Writer 2's fencing WAL with the higher epoch and then consider itself fenced. Really neat that we caught this with fizbee!

Reply:
What if we find the fencing Wal id first, before writing updated manifest? I will update the spec and check. Before that, I want to make sure that spec indeed represents what we intended to do.

Reply:
This is indeed awesome. Hat tip to @jayaprabhakar. I think listing the WALs to the next id _before_ updating the manifest should fix this issue. 🤞

Reply:
The thing we need to ensure is that we check the epoch associated with every WAL SST that was not written by the current writer. So this includes any SSTs returned by the listing, and then any time we fail to write an SST because of a CAS error.

Reply:
So the writer protocol should be:
1. increment the epoch in the manifest to E
2. list the WAL SSTs and write a fencing SST at the next SST ID. If this fails with CAS error, check the epoch in the metadata. If it's lower than E, repeat (2) otherwise exit. 
3. replay the WAL. During replay, check the epoch in every SST. If any have epoch higher than E, exit.

Reply:
Fixing the `next_wal` before writing manifest passes the model checker. 
I am not sure yet if the spec is fully correct though.

Reply:
> The thing we need to ensure is that we check the epoch associated with every WAL SST that was not written by the current writer. So this includes any SSTs returned by the listing, and then any time we fail to write an SST because of a CAS error.

Doesn't bumping the writer epoch guarantee that the only place where a later epoch could occur is at or after the next WAL position?

If so, then getting the current wal position _before_ bumping the epoch should guarantee that the wal position will either be an SST from the current writer (success) or an SST from a future writer (which will fence the current writer).

- Writer 1 lists wals, and sees wal/1.sst as the next slot.
- Writer 1 - creates manifest 1.manifest with epoch 1
- Writer 2 lists wals, and sees wal/1.sst as the next slot.
- Writer 2 - creates manifest 2.manifest with epoch 2
- Writer 2 writes wal/1.sst
- Writer 1 writes to wal/1.sst and CAS fails (fenced)

Or, the last two bullets could be swapped:

- Writer 1 writes wal/1.sst
- Writer 2 writes wal/1.sst and CAS fails (not fenced since epoch is greater)
- Writer 2 tries again at wal/2.sst

Note that I don't think the tablestore.rs currently does the wal advance for SSTs. I think it currently stomps on them--only the manifest is CAS'd at the moment. But the protocol described in https://github.com/slatedb/slatedb/blob/main/rfcs/0001-manifest.md says to advance until an open slot is available.

Reply:
Yeah that also works and seems simpler. We should actually do the check for bullets 2 and 3 and see if the epoch in the WAL SST is actually less than the writer's SST (we don't do this in the current impl), and move on to the next WAL SST if it is. If we don't the protocol will ensure correctness but could result in spuriously detecting that the writer is fenced and exiting.

Reply:
Actually, its still a bit tricky because we'd need to make sure that we include any SSTs that we find when probing for the fencing SST when we do the replay. So the protocol should look like:

1. List WAL SSTs. Lets say that the last seen WAL SST is N.SST and the list of SSTs is L
2. Bump the writer epoch to E
3. Initialize F to N+1
4. Write a fencing SST at F.SST
    5. If it fails with a CAS, then check the epoch in F.SST.
    6. If it is greater than E, exit
    7. If it is less than E:
        8. add F.SST to L
        9. set F to F+1
        10. Go back to 4
11. Replay L

Reply:
Instead of 4.5, we could also read the next manifest. If it was due to a new writer, there must be a manifest present. 
Current implementation does this part already.

Reply:
Ah yeah that's right I always forget that's how we wound up deciding to do it.

Reply:
Me too 😂 but I think this way should work ok. Less data, too.